assayist/client/modify.py

```diff
@@ -0,0 +1,104 @@
+# SPDX-License-Identifier: GPL-3.0+
+
+import itertools
+
+import neomodel
+
+from assayist.common.models.source import Component
+from assayist.client.logging import log
+
+
+@neomodel.db.transaction
+def set_component_names(c_name, c_type, c_namespace='', alternatives=None):
+    """
+    Create, update, and/or merge components to ensure one canonical component remains.
+
+    :param str c_name: the canonical name of the component
+    :param str c_type: the canonical type of the component
+    :param str c_namespace: the canonical namespace of the component
+    :kwarg list alternatives: the alternative names associated with the component
+    :raises ValueError: if the input isn't the proper type or c_name's value is also in alternatives
+    """
+    for arg in (c_name, c_type, c_namespace):
+        if not isinstance(arg, str):
+            raise ValueError('c_name, c_type, and c_namespace must be strings')
+
+    for arg in (c_name, c_type):
+        if arg == '':
+            raise ValueError('c_name and c_type cannot be empty')
+
+    if alternatives is None:
+        alternatives = []
+    if type(alternatives) not in (list, tuple, set):
+        raise ValueError('The alternatives keyword argument must be None or a list, tuple, or set')
+
+    for alternative in alternatives:
+        if not isinstance(alternative, str):
+            raise ValueError('The alternatives keyword argument must only contain strings')
+
+    if c_name in alternatives:
+        raise ValueError('The canonical name can\'t also be an alternative')
+
+    # Create a WHERE clause that checks to see if there is a component with the canonical name or
+    # alternative name with the passed in canonical name and alternatives
+    names_where_clause = [
+        'c.canonical_name = "{0}" OR "{0}" in c.alternative_names'.format(name)
+        for name in itertools.chain([c_name], alternatives)
+    ]
+    query = """
+    MATCH (c:Component {{canonical_type: "{}", canonical_namespace: "{}" }})
+    WHERE {}
+    RETURN c
+    """.format(c_type, c_namespace, ' OR '.join(names_where_clause))
+
+    results, _ = neomodel.db.cypher_query(query)
+    components = [Component.inflate(row[0]) for row in results]
```

Comment:
As a personal preference, I would prefer if the query building and inflating logic lived in a Component class method.